### ISSUE TITLE
Clarification regarding the noise term during prediction

### ISSUE THREAD

**Issue by jonathanbreitg** [Sid3503/NoProp#4]
Looking at the code for the de-noising inference step in the Notebook:
```python
# Inference (denoising)
def predict(x):
    z_t = torch.randn(1, embed_dim)  # Start from noise
    x_features = cnn(x.unsqueeze(0))
    for t in reversed(range(T)):
        u_hat = mlps[t](x_features, z_t)
        z_t = torch.sqrt(alpha[t]) * u_hat + torch.sqrt(1 - alpha[t]) * torch.randn_like(u_hat)
    return torch.argmax(z_t)  # Final prediction
```
when computing z_t you use random noise epsilon: $$z_t = \sqrt{\alpha_t}\hat{u} + \sqrt{1-\alpha_t}\epsilon $$, where $$\epsilon \in \mathcal{N}_d(\epsilon | 0,1)$$ (Equation 6[?] in the arXiv paper using the reparameterization trick for Gaussians). Just like in standard generative Diffusion models we use noise to accurately capture the distribution (Langevin sampling), However surprisingly simply commenting out the noise term naively seems to increase the accuracy of the model, especially earlier in training. 

This is the accuracy i got using the unmodified prediction code on the MNIST dataset (changing the optimizer to AdamW, adding weight_decay=0.001) from the notebook (vanilla) and the noiseless modified version while running for 100 epochs, testing the accuracy of the model on the entire 10K testing dataset after each epoch:
<img width="640" height="480" alt="Image" src="https://github.com/user-attachments/assets/c3bf2a10-357c-49b8-8839-91a3da331758" />

This is of course worse performance than as reported in the paper using the same setup for NoProp-DT:
<img width="892" height="152" alt="Image" src="https://github.com/user-attachments/assets/bdb3e179-691f-4d7e-b10a-3a87f32b3292" />
What do i need to change in order to replicate the standard results in the paper?

Also where is the hyper-parameter $$\eta$$ in the definition of the objective used in the algorithm? It doesn't appear in the notebook as far as i can tell, since we simply compute the L2 loss for each block independently

given that this is a fairly low dimensional toy problem, it's likely that this doesn't generalize and in fact probably hurts the accuracy on more difficult problems (CIFAR-100 For example, which i'll test later ), still , why would removing the noise term increase the Accuracy that dramatically in this case? Maybe because this is a classifier accurately capturing the posterior is less important than simply naively following the gradient. Or perhaps the relatively low dimensionality of the problem helps somehow. I'd like your input!

**Comment by Sid3503** [comment]
Hey! Thanks a ton for diving into this so deeply — this is a great question and you’re absolutely right to notice this detail. Let me clarify a few things for you:

**1. Why does removing the noise sometimes *improve* accuracy?**
This actually makes sense for MNIST. In diffusion models, the final noise term during inference is there to properly *sample* from the learned distribution — you see this in generative tasks where diversity matters (like generating images). But for classification, you usually just want the single most likely label.

So when you *remove* the noise, you’re making the denoising fully deterministic — you’re always picking the mean prediction instead of adding randomness on top. For simple, low-dimensional data like MNIST, this can make the prediction more confident and stable. So ironically, you might see “better” accuracy, because the randomness isn’t helping much in this setting — it’s great for generation but not needed for just picking the top class.


**2. But the paper still uses it — why?**
Good point! The paper keeps the noise because it sticks to the full *diffusion process* idea. The whole goal is to match the theory: the denoiser learns to reverse the forward noising step in a fully probabilistic way. This is important for sampling and modeling the true posterior.
So it’s more about staying faithful to the diffusion math than about squeezing out the absolute best classification score on a toy dataset.


**3. Where’s `η`?**
Good catch — in this notebook, `η` isn’t exposed directly. In the paper’s equations, `η` controls the scale of the noise in the SDE. In practice, it ends up baked into your noise term like this:

```python
z_t = torch.sqrt(alpha[t]) * u_hat + torch.sqrt(1 - alpha[t]) * torch.randn_like(u_hat)
```

To match the paper exactly, you could make `η` explicit, like this:

```python
z_t = torch.sqrt(alpha[t]) * u_hat + torch.sqrt(1 - alpha[t]) * torch.randn_like(u_hat) * eta
```

and just tune `eta` (which is usually 1.0 by default). This gives you control over how strong the injected noise is.

**In short:**

* For *pure classification*, skipping the noise can help because you’re picking the “best guess” directly.
* For *proper diffusion sampling*, you’d want to keep the noise — especially on more complex tasks (like CIFAR-100).
* Feel free to tweak `η` and see how it affects things!


**One more thing:**
The main goal of this repo is to show a *walkthrough* of the NoProp paper’s core idea — not to chase SOTA accuracy. It’s mainly here to help people understand how the forward diffusion and denoising work step by step in code. So there’s a tradeoff between sticking close to the paper’s math vs hacking it for maximum accuracy — but you can explore both!


This is a really good insight — thanks again for sharing your experiments and plots. If you test this on a bigger dataset, definitely share how the noisy vs. noiseless version compares — I’d love to see it!
